contracts/modules/legacy/LegacyPooledStaking.sol

```diff
@@ -1197,208 +1198,196 @@ contract LegacyPooledStaking is IPooledStaking, MasterAwareV2 {
-      address oldProductId = stakers[stakerAddress].contracts[i];
       uint productId;
-      try productsV1.getNewProductId(oldProductId) returns (uint v) {
-        productId = v;
+      try productsV1.getNewProductId(productAddress) returns (uint id) {
+        productId = id;
       } catch {
-        emit ProductNotFound(oldProductId);
+        emit ProductNotFound(productAddress);
         continue;
       }
+
       products[i] = productId;
-      stakes[i] = min(stakers[stakerAddress].stakes[oldProductId], deposit);
-      migratableCount++;
+      stakes[i] = min(stakers[stakerAddress].stakes[productAddress], deposit);
+      productsToBeMigratedCount++;
     }
 
-    params = new ProductInitializationParams[](migratableCount);
-    uint migrateAtIndex = 0;
+    productInitParams = new ProductInitializationParams[](productsToBeMigratedCount);
 
-
-    for (uint i = 0; i < contractCount; i++) {
+    uint index = 0;
+    for (uint i = 0; i < stakedProductsCount; i++) {
       if (stakes[i] == 0) {
         continue;
       }
+
       uint96 price = getV1PriceForProduct(products[i]);
       if (price == type(uint96).max) {
         continue;
       }
-      params[migrateAtIndex] = ProductInitializationParams(
+
+      productInitParams[index] = ProductInitializationParams(
         products[i], // productId
         uint8(min(stakes[i] * 1e18 / deposit / 1e16, 100)), // weight (0-100)
         price / 1e16, // initialPrice with a 100_00 denominator
         price / 1e16 // targetPrice with a 100_00 denominator
       );
-      migrateAtIndex++;
+      index++;
     }
   }
 
+  function migrateToPool(
+    StakingPoolMigrationData memory migrationData,
+    ProductInitializationParams[] memory productInitParams
+  ) internal {
+    ( /* uint stakingPoolId */, address stakingPoolAddress) = cover.createStakingPool(
+      migrationData.managerAddress,
+      migrationData.isPrivatePool,
+      migrationData.initialPoolFee,
+      migrationData.maxPoolFee,
+      productInitParams,
+      migrationData.ipfsDescriptionHash
+    );
 
-  struct StakingPoolMigrationData {
-    uint initialPoolFee;
-    uint maxPoolFee;
-    uint[TRANCHE_COUNT] stakerTrancheRatios;
-    uint deposit;
-    address stakerAddress;
-    bool isPrivatePool;
-    string ipfsDescriptionHash;
-    address managerAddress;
+    // TODO - is this ok here or do we want granular approvals in the loop below?
+    token().approve(address(tokenController()), migrationData.deposit);
+
+    uint totalStakeRatio = 0;
+    uint totalStake = 0;
+    uint firstTrancheId = block.timestamp / 91 days + 1;
+
+    for (uint i = 0; i < MAX_ACTIVE_TRANCHES; i++) {
+      uint trancheDeposit = migrationData.deposit * migrationData.trancheStakeRatio[i] / 100;
+      if (trancheDeposit == 0) {
+        continue;
+      }
+
+      totalStakeRatio += migrationData.trancheStakeRatio[i];
+      // Prevent dust due to precision loss and deposit everything remaining in the last tranche
```

Comment:
open to feedback here

Reply:
i think it's fine, the code runs only once we don't need to optimize it too much

Allocating all to last tranche that's non-zero works